examples/performant_matmul.jl

```diff
@@ -0,0 +1,82 @@
+using KernelAbstractions
+using StaticArrays
+using Test
+
+const TILE_DIM = 32
+
+@kernel function coalesced_matmul_kernel!(output, @Const(input1), @Const(input2), N, R, M,
+                                             ::Val{BANK}=Val(1)) where BANK
+     gi, gj = @index(Group, NTuple)
+     i, j   = @index(Local, NTuple)
+
+     TILE_DIM = @uniform groupsize()[1]
+
+     # +1 to avoid bank conflicts on shared memory
+     tile1 = @localmem eltype(output) (TILE_DIM+BANK, TILE_DIM)
+     tile2 = @localmem eltype(output) (TILE_DIM+BANK, TILE_DIM)
+
+     # private variable for tile output
+     outval = @private eltype(output) 1
+     @inbounds outval[1] = -zero(eltype(output))
+
+     @uniform N = size(output, 1)
+     # number of tiles depends on inner dimension
+     @uniform NUM_TILES = div(R + TILE_DIM - 1, TILE_DIM)
+
+     # loop over all tiles needed for this calculation
+     for t in 0:NUM_TILES-1
+         # Can't use @index(Global), because we use a smaller ndrange
+         I = (gi-1) * TILE_DIM + i
```

Comment:
I'm sure there is a reason for this, but I'm not sure I understand it. Is there an obvious reason why `I = (gi-1) * TILE_DIM + i` here could actually give a different result than `I = (gi-1) * TILE_DIM + i` a few lines below? Is the idea that one should always recompute `I, J` before using them because `i, gi, j, gj` may have changed?

Reply:
It needs to be recalculated after the `@synchronize` otherwise you run into scope issues. This is the kind of problem KernelCompiler.jl will hopefully be able to solve.